compiler/src/template-api/Table.js

```diff
@@ -0,0 +1,318 @@
+            );
+        if (typeof args[requiredArgs[i]] !== requiredArgsTypes[i])
+            throw new Error(
+                "Constructing Table: " +
+                    requiredArgs[i] +
+                    " must be " +
+                    requiredArgsTypes[i] +
+                    "."
+            );
+        if (requiredArgsTypes[i] == "string")
+            if (args[requiredArgs[i]].length == 0)
+                throw new Error(
+                    "Constructing Table: " +
+                        requiredArgs[i] +
+                        " cannot be an empty string."
+                );
+    }
+
+    var table_name = "table_" + args.name;
+
+    var transform_func = getTableTransformFunc(table_name);
+
+    // schema and query first for buiding layer
+    var schema = args.fields.concat(["rn", "y"]);
+
+    var query = args.query || genQuery();
+
+    var transform = new Transform(query, args.db, transform_func, schema, true);
+
+    Layer.call(this, transform, false);
+
+    // TODO: check the type of arguments
+    this.cell_h = args.cell_h || 40;
+    this.x = args.x || 0;
+    this.y = args.y || 0;
+    this.db = args.db;
+
+    var sum_width = 0;
+    var centroid_x = 0;
+    if (typeof args.width === "number") {
+        sum_width = args.width;
+        centroid_x = this.x + args.width / 2;
+    } else if (typeof args.width === "object") {
+        for (key in args.width) {
+            if (typeof args.width[key] !== "number") {
+                throw new Error(
+                    "Error, \
+                    table.args.width with non-numeric object"
+                );
+            }
+            sum_width += args.width[key];
+        }
+        centroid_x = this.x + sum_width / 2;
+    } else {
+        console.log("DEFAULT WIDTH");
+        sum_width = args.fields.length * 100;
+        centroid_x = this.x + sum_width / 2;
+    }
+    this.width = sum_width;
+
+    var th_args = args.heads;
+    if (!th_args) {
+        this.heads = {
+            th_h: 0,
+            names: []
+        };
+    } else if (th_args == "auto") {
+        this.heads = {
+            th_h: this.cell_h,
+            names: args.fields
+        };
+    } else if (!("th_h" in th_args || "names" in th_args)) {
+        throw new Error("in complete heading definition");
+    } else if (typeof th_args.th_h !== "number") {
+        throw new Error("heading height must be number");
+    } else if (
+        typeof th_args.names !== "object" ||
+        th_args.names.length != args.fields.length
+    ) {
+        throw new Error("fields and heads length not equal");
+    } else if (th_args.th_h <= 0) {
+        throw new Error("heading height must be positive");
+    } else {
+        this.heads = th_args;
+    }
+
+    var tableRenderingParams = {
+        // table_name: {
+        [table_name]: {
+            x: this.x,
+            y: this.y,
+            heads: this.heads,
```

Comment:
I'm a bit confused by `heads` here. Seems like `heads` is simply a dictionary containing both `cell_h` and `fields`?

Reply:
user can rename fields here, or make table headers taller or shorter than common lines